### PR TITLE
rearrange adjoint parameter order

### PR DESCRIPTION
**Pull request by marcrasi**
This changes the adjoint parameter order from `(originalParameters..., checkpointsStruct?, originalResult, seed)` to `(seed, checkpointsStruct?, originalResult, originalParameters...)`.

The motivation is that this will let us form a JVP by partially applying the adjoint to `(checkpointsStruct?, originalResult, originalParameters...)`.

Of course, this will break any existing user code that defines custom adjoints. Think I should warn on some mailing lists?

### CODE REVIEW

File: lib/SILOptimizer/Mandatory/TFDifferentiation.cpp [hunk 1]
```diff
@@ -2702,24 +2702,20 @@ class AdjointGen {
...
   SmallVector<SILResultInfo, 8> adjResults;
   auto origTy = original->getLoweredFunctionType();
   auto origParams = origTy->getParameters();
-  assert((!origTy->hasSelfParam() ||
-          origParams.back() == origTy->getSelfParameter()) &&
-         "self argument should be the last one");
-  auto origParamsWithoutSelf = origTy->hasSelfParam() ?
-      origParams.slice(0, origParams.size() - 1) : origParams;
-
-  // Add adjoint parameters for the original non-self parameters. If there is a
```
Comment (marcrasi) [inline lib/SILOptimizer/Mandatory/TFDifferentiation.cpp]: Excitingly, this new order simplifies a lot of the method stuff!

File: lib/Sema/TypeCheckAttr.cpp [hunk 2]
```diff
@@ -2180,6 +2180,9 @@ static FuncDecl *resolveAutoDiffAssociatedFunction(
       overloadDiagnostic, ambiguousDiagnostic, notFunctionDiagnostic,
       lookupOptions, hasValidTypeContext, invalidTypeContextDiagnostic);
 
+  if (!candidate)
```
Comment (marcrasi) [inline lib/Sema/TypeCheckAttr.cpp]: I noticed this bug while working on this, so I slipped this fix in.

**Comment by marcrasi** [comment]
@swift-ci please test tensorflow

**Comment by rxwei** [comment]
Exciting progress! Instead of reversing the parameter order of user-defined primitives, I think we can simplify this even further by dropping the notion of primal and adjoint from user syntax altogether. We’ll require all users to specify VJP or VJP. That way, JVP and VJP functions will already return a partially applied function, and it’d be much closer to the final syntax design.

**Comment by marcrasi** [comment]
@swift-ci please test tensorflow

**Comment by marcrasi** [comment]
@swift-ci please test tensorflow

**Comment by marcrasi** [comment]
@swift-ci please test tensorflow

**Comment by marcrasi** [comment]
I do think we should remove primal and adjoint from user syntax, and replace them with jvp/vjp. I think that the core ad pass will need to be able to synthesize and use jvp/vjp before I can do that though. So I propose doing this PR first to unblock jvp/jvp synthesis & use, then doing jvp/vjp synthesis & use, and then finally removing primal and adjoint from user syntax.

**Comment by marcrasi** [comment]
@swift-ci please test tensorflow

**Review by rxwei — APPROVED** [review]

**Comment by marcrasi** [comment]
@swift-ci please test tensorflow

**Pull request merged**